### ISSUE TITLE
add shuffling option to tf.train.batch which handles variable-length sequence

### ISSUE THREAD

**Issue by jinghuangzhu** [tensorflow/tensorflow#2354]
tf.train.batch is the only one can handle variable-length sequence, but it does not have shuffling option.
For RNN training, shuffling is needed.

**Comment by girving** [comment]
Use `tf.shuffle_batch`.

**Comment by jinghuangzhu** [comment]
However, tf.train.shuffle_batch does not offer padding zero to variable-length sequences, which is needed to deal with these kind of data for RNN training.

**Comment by girving** [comment]
@ebrevdo: Is there a way to get both bits of functionality in the same routine currently?  If not, what would be the best way to support it?

**Comment by ebrevdo** [comment]
Right now, batch and batch_join use a PaddingFIFOQueue if dynamic_pad=True otherwise a FIFOQueue.  shuffle_batch uses a ShuffleQueue which does not do batchwise padding.

Maybe by combining the PaddingFIFOQueue with the ShuffleQueue underneath if someone passes dynamic_pad=True to shuffle_batch (which right now doesn't have that parameter)?  Marking as contributions welcome.  I don't have bandwidth to work on this in the near future.

**Comment by ibab** [comment]
@ebrevdo: I've had a look at this and realized that the output shape of `PaddingFIFOQueue` is still dynamic where the input shape was dynamic. This makes sense, as we can only find out what the maximum size per dimension is when we execute `dequeue_many` at runtime.
But the dynamic output shape prevents me from piping the padded tensors into a `ShuffleQueue`.
So I'm sure if it's possible to implement this without making a new C++ op or adjusting the existing ones.

**Comment by petrux** [comment]
I have spent the last week and maybe more dealing with such [issue](http://stackoverflow.com/questions/43367697/batching-and-shuffling-padded-tf-train-sequenceexample). So, help me to uderstand (and save a lot of time): can we state that currently there is _no way_ of shuffling and batching using the official TF input pipeline utilities?

**Comment by ebrevdo** [comment]
You can easily shuffle and batch using tf.train.shuffle_batch *so long as
you don't have variable-length input tensors*.

**Comment by petrux** [comment]
Hi @ebrevdo and thanks for your reply. Actually I came out with this idea:
* feed my parsed protobufs into a `tf.PaddingFIFOQueue`
* feed it into a `tf.RandomShuffleQueue`
and apparently it works fine. So, here follows the code of my test example (which is quite ugly, but no big deal). Hopefully I will set up something more interesting in one of my projects and maybe share the link here. Any comment is _more_ than welcome.

```python
import random

import tensorflow as tf


MIN_LEN = 6
MAX_LEN = 12
NUM_EXAMPLES = 10
BATCH_SIZE = 7
PATH = 'ciaone.tfrecords'
MIN_AFTER_DEQUEUE = 10
NUM_THREADS = 2
SAFETY_MARGIN = 1
CAPACITY = MIN_AFTER_DEQUEUE + (NUM_THREADS + SAFETY_MARGIN) * BATCH_SIZE


def generate_example():
    """Generate an example."""
    length = random.randint(MIN_LEN, MAX_LEN)
    input_ = [random.randint(1, 10) for _ in xrange(length)]
    avg = sum([1.0 * item for item in input_]) / len(input_)
    output = [item for item in input_ if item >= avg]
    return input_, output


def encode(input_, output):
    """Encode a pair of input, output tensor into a tf.train.Example."""
    example = tf.train.Example(
        features=tf.train.Features(
            feature={
                'input': tf.train.Feature(
                    int64_list=tf.train.Int64List(
                        value=input_)),
                'output': tf.train.Feature(
                    int64_list=tf.train.Int64List(
                        value=output))}))
    return example


def decode(example):
    """Decode an example into a tuple of input, output tensors."""
    features = {
        'input': tf.VarLenFeature(tf.int64),
        'output': tf.VarLenFeature(tf.int64)
    }
    parsed = tf.parse_single_example(
        serialized=example,
        features=features)
    input_ = parsed['input']
    output = parsed['output']
    input_ = tf.sparse_tensor_to_dense(parsed['input'])
    output = tf.sparse_tensor_to_dense(parsed['output'])
    return input_, output


def main():
    """Run the example."""

    # 1. decode the records from the file.
    file_queue = tf.train.string_input_producer(
        [PATH], shuffle=True, num_epochs=2)
    reader = tf.TFRecordReader()
    key, value = reader.read(file_queue)
    input_, output = decode(value)

    # 2. padding queue.
    padding_queue = tf.PaddingFIFOQueue(
        capacity=CAPACITY,
        dtypes=[tf.string, tf.int64, tf.int64],
        shapes=[[], [None], [None]])
    padding_enqueue_op = padding_queue.enqueue([key, input_, output])
    padding_queue_runner = tf.train.QueueRunner(
        padding_queue, [padding_enqueue_op] * NUM_THREADS)
    tf.train.add_queue_runner(padding_queue_runner)
    padding_dequeue_op = padding_queue.dequeue_up_to(BATCH_SIZE)

    # 3. shuffling queue that tolerates different shapes.
    shuffle_queue = tf.RandomShuffleQueue(
        capacity=CAPACITY,
        min_after_dequeue=MIN_AFTER_DEQUEUE,
        dtypes=[tf.string, tf.int64, tf.int64],
        shapes=None)
    shuffle_enqueue_op = shuffle_queue.enqueue(padding_dequeue_op)
    shuffle_queue_runner = tf.train.QueueRunner(
        shuffle_queue, [shuffle_enqueue_op] * NUM_THREADS)
    tf.train.add_queue_runner(shuffle_queue_runner)
    shuffle_dequeue_op = shuffle_queue.dequeue()
    fetches = shuffle_dequeue_op

    # 4. run.
    with tf.Session() as sess:
        sess.run(tf.local_variables_initializer())
        sess.run(tf.global_variables_initializer())
        coord = tf.train.Coordinator()
        threads = tf.train.start_queue_runners(coord=coord)
        try:
            while True:
                keys, inputs, outputs = tuple(sess.run(fetches))
                for k, i in zip(keys, inputs):
                    print(k + ': ' + str(i))
                print(inputs.shape)
                print
        except tf.errors.OutOfRangeError as ex:
            coord.request_stop(ex=ex)
        finally:
            coord.request_stop()
            coord.join(threads)
    print(' -- done -- ')


if __name__ == '__main__':
    # Before running the example, generate a bunch of examples
    # and save them in a file as TFRecords.
    with tf.python_io.TFRecordWriter(PATH) as writer:
        for _ in xrange(NUM_EXAMPLES):
            ii, oo = generate_example()
            example = encode(ii, oo)
            writer.write(example.SerializeToString())
    main()
```

**Comment by ebrevdo** [comment]
Looks like you are shuffling the *minibatches*, not the individual
minibatch elements, with your shuffle queue (you're calling enqueue and
dequeue, not enqueue_many and dequeue_many).  So if there's heavy
correlation locally between inputs, your minibatches will still have that
correlation and you lose the expected benefits of shuffling.

**Comment by petrux** [comment]
Hi @ebrevdo and thanks for the reply. I was shuffling minibatches indeed. So, to sum everything up, there is no way to do that? Otherwise, could you/anyone suggest a modification to my code to shuffle element across different minibatches? Thanks!  
  
(by the way, I thought it is quite a common scenario...)

**Comment by ebrevdo** [comment]
It is something folks want to do; and it is unfortunate there is no
PaddingShuffleQueue.  There was a PR a while back to add it, but the PR was
withdrawn (perhaps the code is sitting in the original author's git repo
still?).  You can maybe review github issues for "PaddingShuffleQueue" to
see if the code can still be found.

**Comment by petrux** [comment]
Hi @ebrevdo and thanks for the reply. I followed your suggestion and scanned a bit through #5312 and #5147 and [this guy](https://github.com/tensorflow/tensorflow/issues/5147#issuecomment-271086206) made my day.

**Comment by amadupu** [comment]
I have taken some inputs from the above suggestions, however facing some issue which I posted on https://stackoverflow.com/questions/44838356/paddingfifoqueue-enqueue-many-throwing-value-error-shapes-must-be-equal-rank

pls let me know if anyone has some clue on what is that I'm doing wrong

**Comment by alquraishi** [comment]
Just a ping to see if there's been any activity on this? Seems like very basic functionality.

**Comment by mrry** [comment]
The `tf.data` API supports shuffling elements containing variable-sized tensors, using [`tf.data.Dataset.shuffle()`](https://www.tensorflow.org/versions/r1.4/api_docs/python/tf/data/Dataset#shuffle). We're recommending that people use this API for new code, because it tends to be more efficient (and often much simpler!) than code written using `tf.train.batch()` (and related methods).

**Comment by alquraishi** [comment]
Thanks @mrry. Does the new `Dataset` API support automatic bucketing by length? I was hoping to use `tf.contrib.training.bucket_by_sequence_length` and already have everything implemented using the old `tf.train`-based API.

**Comment by ebrevdo** [comment]
Yes; see the NMT tutorial code here
<https://github.com/tensorflow/nmt/blob/master/nmt/utils/iterator_utils.py#L197>
(this
buckets by sequence length + batches with padding).  I believe
tensor2tensor has similar code here
<https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/data_reader.py#L117>
.

**Comment by zh794390558** [comment]
Tanks @mrry @ebrevdo

**Comment by yingwang** [comment]
I also worked on an RNN model. You can shuffle your training data before batching. For instance, when you deserialize data from TFRecord files, you can use tf.train.string_input_producer(fileList, numEpochs, shuffle) to shuffle data within each epoch.
https://www.tensorflow.org/api_docs/python/tf/train/string_input_producer
